File: bench/bench.py
#!/usr/bin/env python3
"""Timing benchmark for the Python DiffusionGeometry tracer-bullet pipeline.

Point it at a checkout of the original Python package via $DIFFUSION_GEOMETRY_PY,
or place that repo alongside this one (../DiffusionGeometry).
"""
import os, sys, time
import numpy as np

_HERE = os.path.dirname(os.path.abspath(__file__))
for cand in (os.environ.get("DIFFUSION_GEOMETRY_PY", ""),
             os.path.abspath(os.path.join(_HERE, "..", "..", "DiffusionGeometry"))):
    if cand and os.path.isdir(cand) and cand not in sys.path:
        sys.path.insert(0, cand)

from diffusion_geometry.core.geometry.diffusion_geometry import DiffusionGeometry

SIZES = [200, 500, 1000, 2000, 3000, 5000]
D, KNN, NFB, NCOEF, REPS = 3, 20, 32, 16, 5


def timeit(f, reps=REPS):
    ts = []
    for _ in range(reps):
        t = time.perf_counter(); f(); ts.append(time.perf_counter() - t)
    return min(ts), float(np.median(ts))


def build(data):
    return DiffusionGeometry.from_point_cloud(
        data, knn_kernel=KNN, n_function_basis=NFB, n_coefficients=NCOEF)


print("stage,n,impl,warm_min_s,warm_median_s,first_call_s")
for n in SIZES:
    rng = np.random.default_rng(0)
    data = rng.standard_normal((n, D))

    t = time.perf_counter(); build(data); first = time.perf_counter() - t
    bmin, bmed = timeit(lambda: build(data))
    print(f"build,{n},python,{bmin:.5f},{bmed:.5f},{first:.5f}")

    t = time.perf_counter()
    build(data).laplacian(0).spectrum(eigvals_only=True)
    sfirst = time.perf_counter() - t
    smin, smed = timeit(lambda: build(data).laplacian(0).spectrum(eigvals_only=True))
    print(f"build+spectrum,{n},python,{smin:.5f},{smed:.5f},{sfirst:.5f}")
    sys.stdout.flush()

File: bench/bench.jl
#!/usr/bin/env julia
# Timing benchmark for the Julia DiffusionGeometryJ tracer-bullet pipeline.
using DiffusionGeometryJ
using Random, Statistics, Printf

const SIZES = [200, 500, 1000, 2000, 3000, 5000]
const D = 3
const KNN = 20
const NFB = 32
const NCOEF = 16
const REPS = 5

# time a thunk `reps` times, return (min, median) in seconds
function timeit(f; reps=REPS)
    ts = Float64[]
    for _ in 1:reps
        push!(ts, @elapsed f())
    end
    (minimum(ts), median(ts))
end

function build(data)
    from_point_cloud(data; knn_kernel=KNN, n_function_basis=NFB, n_coefficients=NCOEF)
end

println("stage,n,impl,warm_min_s,warm_median_s,first_call_s")
for n in SIZES
    Random.seed!(0)
    data = randn(n, D)

    # --- from_point_cloud ---
    t_first = @elapsed build(data)          # includes JIT on first n only
    bmin, bmed = timeit(() -> build(data))
    @printf("build,%d,julia,%.5f,%.5f,%.5f\n", n, bmin, bmed, t_first)

    # --- laplacian(0).spectrum ---
    dg = build(data)
    sp_first = @elapsed spectrum(laplacian(dg, 0); eigvals_only=true)
    smin, smed = timeit(() -> spectrum(laplacian(build(data), 0); eigvals_only=true))
    @printf("build+spectrum,%d,julia,%.5f,%.5f,%.5f\n", n, smin, smed, sp_first)
    flush(stdout)
end
